BootloaderCommonPkg/Library/PagingLib/PagingMap.c

```diff
@@ -44,60 +50,100 @@ MapMemoryRange (
   )
 {
   UINTN          Idx;
+  UINTN          End;
   UINTN          EntryNum;
-  UINTN         *PageTable;
+  UINTN         *PageTable1G;
+  UINTN         *PageTable2M;
   UINTN         *PageTable4K;
-  UINT32         PageBase;
+  UINTN          PageBase;
   UINT32         Attribute;
-  UINT32         Address;
+  UINTN          Address;
   UINT32         PteOffset;
   UINT32         PdeOffset;
-  UINT32         Alignment;
+  UINT32         PdpOffset;
+  UINTN          Alignment;
   UINT32         PageBits;
 
   Attribute    = IA32_PG_P | IA32_PG_RW | IA32_PG_AC;
 
-  // Define CAR region and memory mapping
-  if (((Ranges[0].Start & 0xFFF) != 0) || ((Ranges[0].Limit & 0xFFF) != 0xFFF)) {
+  // Check for Start and Limit PageSize alignment
+  if (((Ranges[0].Start & (Ranges[0].PageSize - 1)) != 0) || ((Ranges[0].Limit & (Ranges[0].PageSize - 1)) != (Ranges[0].PageSize - 1))) {
     return EFI_INVALID_PARAMETER;
   }
 
-  PageBits  = GetPageShiftBits ();
-  if (PageBits == 21) {
-    // 64 bit mode
-    PteOffset = 6 * EFI_PAGE_SIZE;
-    PdeOffset = 2 * EFI_PAGE_SIZE;
-    EntryNum  = 512;
-  } else {
-    // 32 bit mode
-    PteOffset = 1 * EFI_PAGE_SIZE;
-    PdeOffset = 0;
-    EntryNum  = 1024;
-  }
-
-  Alignment = (UINT32) LShiftU64 (1, PageBits) - 1;
-  PageBase  = Ranges[0].Start & ~Alignment;
-  if (((Ranges[0].Start & ~Alignment) != PageBase) || ((Ranges[0].Limit & ~Alignment) != PageBase)) {
+  // Check for Mapping PageSize alignment
+  if ( (Ranges[0].Mapping & (Ranges[0].PageSize - 1)) != 0) {

```

Comment:
I think we miss the IsLongModeEnabled() functionality if we calculate alignment based on pagesize and dont calculate PageBits as here. Isnt this more robust?

Reply:
we should not use IsLongModeEnabled() here since 64bit paging might be required in both 32 bit mode or 64 bit mode. It has to be determined using input parameter.

Reply:
Got it. Aiden is just asking to use the variable in the if condition. I was confusd with the Alignment  that we set again using PageBits to get the PageBase. Agree, at this point, we need to use PageSize input param. Thanks.